app/src/main/java/com/sopt/now/presentation/Login/LoginViewModel.kt

```diff
@@ -0,0 +1,58 @@
+package com.sopt.now.presentation.Login
+
+
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.sopt.now.domain.AuthRepository
+import com.sopt.now.domain.model.AuthData
+import kotlinx.coroutines.launch
+import retrofit2.HttpException
+
+class LoginViewModel(
+    private val authRepository: AuthRepository
+) : ViewModel() {
+    val loginResult: MutableLiveData<Boolean> = MutableLiveData()
+    val errorMessage: MutableLiveData<String> = MutableLiveData()
+    val userId: MutableLiveData<String> = MutableLiveData()
+    fun logIn(authData: AuthData) {
+        viewModelScope.launch{
+            authRepository.logIn(authData)
+            .onSuccess { response ->
+                val userId =  response.headers()["Location"]
+                userId?.let {
+                    this@LoginViewModel.userId.value =it
+                    loginResult.value = true
+                }
```

Comment:
논란이 많은 scope에서 null체크이지만..

현재 userId는 불변 변수입니다
이러한 불변 변수의 null체크는 if else문을 쓰는게 let보다 성능면에서 더욱 효과적입니다.
반대로 Mutable변수의 경우 let으로 null체크하는것이 효과적이긴 해요!

```kotlin
fun process(str: String?) {
    str?.let { /*Do something*/   }
}
```

```java
public final void process(@Nullable String str) {
   if (str != null) {
      boolean var4 = false;
      /*Do something*/
   }
}
```
위의 let을 자바로 decompile할 경우 저러한 코드가 됩니다! 오히려 이상한 변수가 하나 추가되었죠?

이 경우 그냥 if else를 사용하는 것이 더 낫습니다!

```suggestion
                val userId =  response.headers()["Location"]
                if(userId != null) {
                    _userId.value =userId
                    loginResult.value = true
                }
```

이렇게 쓰는 것을 추천드리는 바이지만.. 
JVM 기반 컴파일러에서 이정도의 오버헤드는 최적화하는데 효과적이라고 하네요

Reply:
오홍 이거는 몰랐던 내용이에요! 새롭게 알아가요!

Reply:
저두 몰랐던 내용 새롭게 알아가요!